account_avatax_exemption_base/models/exemption.py

```diff
@@ -0,0 +1,179 @@
+            ("draft", "Draft"),
+            ("progress", "In Progress"),
+            ("done", "Done"),
+            ("cancel", "Cancelled"),
+        ],
+        default="draft",
+    )
+
+    def _get_document_folder(self):
+        return self.env.company.documents_exemption_folder
+
+    def _check_create_documents(self):
+        return (
+            self.env.company.documents_exemption_settings
+            and super()._check_create_documents()
+        )
+
+    def name_get(self):
+        res = []
+        for record in self:
+            if record.exemption_number:
+                name = f"{record.exemption_number} - {record.partner_id.display_name}"
+            else:
+                name = record.partner_id.display_name
+            if record.exemption_type:
+                name = f"{record.exemption_type.name} - {name}"
+            res.append((record.id, name))
+        return res
+
+    # TODO: Need to check, avalara.salestax not found and also field.
+    # @api.onchange("partner_id")
+    # def onchange_partner_id(self):
+    #     avalara_salestax = (
+    #         self.env["avalara.salestax"]
+    #         .sudo()
+    #         .search([("exemption_export", "=", True)], limit=1)
+    #     )
+    #     if avalara_salestax.use_commercial_entity:
+    #         self.partner_id = self.partner_id.commercial_partner_id.id
+    #         return {"domain": {"partner_id": [("parent_id", "=", False)]}}
+
+    @api.onchange("exemption_type", "group_of_state")
+    def onchange_exemption_type(self):
+        self.business_type = self.exemption_type.business_type.id
+        if self.exemption_type.group_of_state and not self.group_of_state:
+            self.group_of_state = self.exemption_type.group_of_state.id
+        if self.exemption_type or self.group_of_state:
+            state_ids = []
+            if self.exemption_type.group_of_state.state_ids:
+                state_ids += self.exemption_type.group_of_state.state_ids.ids
+            if self.exemption_type.state_ids:
+                state_ids += self.exemption_type.state_ids.ids
+            if self.group_of_state.state_ids:
+                state_ids += self.group_of_state.state_ids.ids
+            self.state_ids = [(6, 0, list(set(state_ids)))]
+
+    @api.onchange("exemption_type", "effective_date")
+    def onchange_effective_date(self):
+        if self.exemption_type.exemption_validity_duration and self.effective_date:
+            self.expiry_date = self.effective_date + timedelta(
+                days=self.exemption_type.exemption_validity_duration
+            )
+
+    @api.onchange("state_ids")
+    def onchange_state_ids(self):
+        if not any(self.exemption_line_ids.mapped("avatax_id")) and not any(
+            self.exemption_line_ids.mapped("add_exemption_number")
+        ):
+            self.exemption_line_ids = [(6, 0, [])]
+        for state_id in self.state_ids.ids:
+            if state_id not in self.exemption_line_ids.mapped("state_id").ids:
```

Comment:
```suggestion
        existing_state_ids = self.exemption_line_ids.mapped("state_id").ids
        for state_id in self.state_ids.ids:
            if state_id not in existing_state_ids:
```
avoid accesses in a loop